core/src/main/java/jenkins/install/StartupUtil.java

```diff
@@ -0,0 +1,153 @@
+
+    /**
+     * Get the type of "startup" currently under way in Jenkins.
+     * @return The type of "startup" currently under way in Jenkins.
+     */
+    public static StartupType getStartupType() {
+        String lastRunVersion = getLastExecVersion();
+
+        // Neither the top level config or the lastExecVersionFile have a version
+        // stored in them, which means it's a new install.
+        if (lastRunVersion.equals("1.0")) {
+            return StartupType.NEW;
+        }
+
+        // We have a last version.
+
+        String currentRunVersion = getCurrentExecVersion();
+        if (lastRunVersion.equals(currentRunVersion)) {
+            return StartupType.RESTART;
+        } else {
+            return StartupType.UPGRADE;
+        }
+    }
+
+    /**
+     * Save the current Jenkins instance version as the last executed version.
+     * <p>
+     * This state information is required in order to determine whether or not the Jenkins instance
+     * is just restarting, or is being upgraded from an earlier version.
+     */
+    public static void saveLastExecVersion() {
+        if (Jenkins.VERSION.equals(Jenkins.UNCOMPUTED_VERSION)) {
+            // This should never happen!! Only adding this check in case someone moves the call to this method to the wrong place.
+            throw new IllegalStateException("Unexpected call to StartupUtil.saveLastExecVersion(). Jenkins.VERSION has not been initialized. Call computeVersion() first.");
+        }
+        saveLastExecVersion(Jenkins.VERSION);
+    }
+
+    /**
+     * Get the last saved Jenkins instance version.
+     * @return The last saved Jenkins instance version.
+     * @see #saveLastExecVersion()
+     */
+    public static @Nonnull String getLastExecVersion() {
+        File lastExecVersionFile = getLastExecVersionFile();
+        if (lastExecVersionFile.exists()) {
+            try {
+                return FileUtils.readFileToString(lastExecVersionFile);
+            } catch (IOException e) {
+                throw new IllegalStateException("Unexpected Error. Unable to read " + lastExecVersionFile.getAbsolutePath(), e);
+            }
+        } else {
+            // Backward compatibility. Use the last version stored in the top level config.xml.
+            VersionNumber storedVersion = Jenkins.getStoredVersion();
+            if (storedVersion != null) {
+                return storedVersion.toString();
+            } else {
+                return "1.0";
+            }
+        }
+    }
+
+    /**
+     * Has "this" Jenkins instance gone through a startup sequence using a version released since the unbundling epoc
+     * i.e. the time when Jenkins stopped pre-bundling unessential plugins.
+     * <p>
+     * This check can be used to determine what kind of upgrade needs to be performed i.e.
+     * <ul>
+     *     <li>Pre unbundling epoc version upgrade: In this case, all plugins ever detached from Jenkins need to be installed.</li>
```

Comment:
@daniel-beck ok thanks. I'm obviously missing something or not asking the right questions.

Here are the 2 upgrade scenarios as I see them...

__Scenario 1__:
Upgrading from an old'ish Jenkins version (that has bundled plugins) e.g. 1.609. Which of the previously bundle plugins (that are no longer bundled) should be auto-installed? I'd have thought you'd need to auto-install them all because you don't know what they've been using.

__Scenario 2__:
Upgrading from a Jenkins version that was released after we unbundled all plugins e.g. upgrading from 1.700 to 1.710 (i.e. some time in the future). So in this situation, they've already gone though at least one upgrade sequence. Which of the previously bundle plugins (that are no longer bundled) should be auto-installed? I'd have thought this is a different upgrade and only plugins that have been detached since 1.700 would need to be auto-install. You can't just install them all again coz you might end up reinstalling something that they uninstalled after the last upgrade (to 1.700).

Where am I wrong there?